Add xplr.util.lscolor and xplr.util.paint

noahmayr opened the pull request:
Adds two of the utility functions discussed in #562 and #564

Please note that this was my first time using Rust, so my code (and code style) is most likely far from perfect.

I'll try to attach an example init.lua file to showcase this a bit better

sayanarijit commented:
Thanks for the pr. I'll review and merge it before the next release (about a month). If you're working on plugins that depend on these apis, try to add lua fallbacks, they don't have to work as long as it doesn't break the actual functionality. Also we can't expect everyone to upgrade xplr as soon as it releases, so fallbacks will be needed.

sayanarijit commented on src/lua/util.rs:
````diff
@@ -317,3 +321,50 @@ pub fn to_yaml<'a>(util: Table<'a>, lua: &Lua) -> Result<Table<'a>> {
+///
+/// Example:
+///
+/// ```lua
+/// xplr.util.lscolor("Desktop")
+/// -- { fg = "Red", bg = nil, add_modifiers = {}, sub_modifiers = {} }
+/// ```
+pub fn lscolor<'a>(util: Table<'a>, lua: &Lua) -> Result<Table<'a>> {
+    let lscolors = LsColors::from_env().unwrap_or_default();
+    let func = lua.create_function(move |lua, path: String| {
+        let style = lscolors
+            .style_for_path(path)
+            .map_or(Style::default(), Style::from);
+        lua::serialize(lua, &style).map_err(LuaError::custom)
+    })?;
+    util.set("lscolor", func)?;
+    Ok(util)
+}
````
Since `style` is capable of handling `nil` as input, I'd expect it to return `nil` instead of `Style::default()` to avoid unnecessary serialization.

sayanarijit commented on src/lua/util.rs:
````diff
@@ -317,3 +321,50 @@ pub fn to_yaml<'a>(util: Table<'a>, lua: &Lua) -> Result<Table<'a>> {
+    Ok(util)
+}
+
+/// Format a string using a style object
+///
+/// Type: function( string, style ) -> string
+///
+/// Example:
+///
+/// ```lua
+/// xplr.util.style("Desktop", { fg = "Red", bg = nil, add_modifiers = {}, sub_modifiers = {} })
+/// -- "\u001b[31mDesktop\u001b[0m"
+/// ```
+pub fn style<'a>(util: Table<'a>, lua: &Lua) -> Result<Table<'a>> {
+    let func =
+        lua.create_function(|lua, (string, style): (String, Option<Table>)| {
+            let style: Style = if let Some(s) = style {
+                lua.from_value(Value::Table(s))?
+            } else {
+                Style::default()
+            };
+            let ansi_style: nu_ansi_term::Style = style.into();
+            Ok::<String, LuaError>(ansi_style.paint(string).to_string())
+        })?;
+    util.set("style", func)?;
+    Ok(util)
+}
````
Let's also handle NO_COLOR. `use crate::ui::NO_COLOR`. Also, instead of the name `style`, I think `paint` sounds better.

sayanarijit replied:
Also, let's avod unnecessary computation (i.e. return string) if style is `nil`.

noahmayr replied:
One thing I wasn't quite sure about, should we return the input string early in case of NO_COLOR to maximize performance or still try to parse the input into a Style object which could potentially fail to consistently give the same error with the same input?

noahmayr replied:
Yeah, that's also what I went with now. Just wanted to make sure

sayanarijit reviewed (changes requested):
Nicely done. Minor changes suggested.

sayanarijit reviewed (approved):
Will be included in the next release.

sayanarijit commented:
How about we actually use it in the default init.lua?

noahmayr commented:
> How about we actually use it in the default init.lua?

Sure, I simplified the permission bit renderer and added file/directory name coloring based on lscolor (ideally use https://github.com/sharkdp/vivid to set a theme you like to test this)

sayanarijit commented:
I have created `dev` branch so that we can merge pulls without affecting the main branch docs. Closing this w.r.t. https://github.com/sayanarijit/xplr/pull/569
